### PR TITLE
XOR

### PR DESCRIPTION
**Pull request by mrwatts88**
Add the ability to evaluate rules composed with XOR, and the helper function, similar to and, or, not.

### CODE REVIEW

File: src/impl.spec.js [hunk 1]
```diff
@@ -20,6 +20,10 @@ test('Implement Regent without init()', (assert) => {
   const SHOULD_WEAR_COAT = or(IS_RAINING, IS_SNOWING, IS_COLD);
 
   assert.true(evaluate(SHOULD_WEAR_COAT, data), 'Should Wear Coat should return true');
+
+  const WET_BUT_NOT_TOO_WET = xor(IS_RAINING, IS_SNOWING);
```
Comment (eisenivan) [inline src/impl.spec.js]: Love the example!

File: src/index.js [hunk 2]
```diff
@@ -25,6 +25,11 @@ export const or = (...rules) => ({
   rules,
 });
 
+export const xor = (...rules) => ({
```
Comment (eisenivan) [inline src/index.js]: We will probably want a test or two making sure that this function exists, and that checks it's param signature. So something like 

```
assert.equal(typeof xor, 'function')
```

File: src/private/parse-composed.js [hunk 3]
```diff
@@ -9,17 +9,21 @@ export default (obj, data, custom = {}) => {
     result = !evaluateRule(notRule, data, custom);
   } else {
     const action = obj.compose;
+    const fxn = rule => (
```
Comment (eisenivan) [inline src/private/parse-composed.js:12]: This is great too! Nice abstraction.

**Review by eisenivan — CHANGES_REQUESTED** [review]
Thanks so much for the PR. We're taking a look at it and will be in touch, but in the mean time I left a few notes.

Could you add unit tests for `index.js`, and `parse-composed.js`? There are `.spec.js` files as siblings to both with a bunch of examples.

How did you stumble on regent? Hope you don't mind me asking.

**Comment by mrwatts88** [comment]
I will look into those changes today. I was browsing the Milwaukee Slack channel and I'll be interning at NWM this Summer so Regent caught my eye.

**Comment by eisenivan** [comment]
Awesome! We are also considering caping the number of arguments `xor` gets to 2. We were initially thinking that it should take `n` arguments, and confirm that some but not all are true, like you implemented, but now we think it would be more clear if it took exactly 2 and threw an error if you passed it more.

Do you have any thoughts on that? I'd love to get some outside thinking on it.

This Wikipedia entry sort of illustrates our thinking. Thanks to @aLittleTooQuiet for digging that up.

https://en.wikipedia.org/wiki/XOR_gate#More_than_two_inputs

**Comment by mrwatts88** [comment]
It's hard to say, but the way I've implemented it isn't typically the way XOR works with multiple inputs. In a digital logic sense, it typically outputs true if the number of true inputs is odd, false otherwise. So there are really 2 options for how to implement XOR with 3 or more arguments. Restricting the number of args to 2 will eliminate that ambiguity.

Tough call, but anyways I have added the tests you asked for. Let me know if you want to make the XOR change.

**Comment by eisenivan** [comment]
@mrwatts88 this looks great! Only tests I'd like to see added would be sad path tests, so basically ensuring it throws an error when called with 0 rules, 1 rule, 3+ rules.

**Comment by mrwatts88** [comment]
@eisenivan okay I've added those tests. Let me know if they aren't in the correct place or if there is anything you need me to change.

**Review by eisenivan — APPROVED** [review]

**Pull request merged**